### PR TITLE
feat(plugins): allow to import separate plugins

### PR DESCRIPTION
**Pull request by Johann-S**
Finally I succeed to achieve that !

With that PR, folks will be able to choose which plugins they want to import and not all of our plugins.

**Examples:**

*Importing Util*
```js
import $ from 'jquery';
import Util from 'bootstrap/js/dist/util';
```

*Importing Popover*
```js
import $ from 'jquery';
import 'popper.js';
import 'bootstrap/js/dist/tooltip';
import 'bootstrap/js/dist/popover';
```

This PR would help a lot in that case: https://github.com/symfony/webpack-encore/issues/338

**TODO**
- [x] Update our docs

Fixes: #26757

### CODE REVIEW

File: build/build-plugins.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    // Do not bundle Util in plugins
+    if (bsPlugins[pluginKey] !== bsPlugins.Util) {
+      external.push(bsPlugins.Util)
+      globals['./util'] = 'Util'
+    }
+
+    // Do not bundle Tooltip in Popover
+    if (bsPlugins[pluginKey] === bsPlugins.Popover) {
```
Comment (stof) [inline build/build-plugins.js]: woudln't `pluginKey === 'Popover'` be simpler ?

**Comment by Johann-S** [comment]
Finally I won't remove `js/dist` files because it should be a breaking change but I'll update our docs to stop mention them

File: build/build-plugins.js (hunk 1 above)
Comment (Johann-S) [reply build/build-plugins.js]: Yep it'll clearer thanks @stof 👍

**Comment by Johann-S** [comment]
ping @XhmikosR

**Comment by Johann-S** [comment]
ping @XhmikosR

File: build/build-plugins.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
+/*!
+ * Script to build our plugins to use them separately.
+ * Copyright 2017-2018 The Bootstrap Authors
```
Comment (XhmikosR) [inline build/build-plugins.js]: nitpicking, but this file is new, so 2018 is enough.

**Comment by XhmikosR** [comment]
1. Isn't there any way to not add yet another set of generated files to the repo?
2. Shouldn't we test these files too?

**Comment by Johann-S** [comment]
Yep I'm not confortable with adding a new set of dist files... I thought about deleting our `js/dist` folder but it can be helpful for those who don't use bundlers

Those files don't have to be tested because the logic is the same.

Bundlers are used a lot nowaday, so we should be able to provide distinct plugins

**Review by XhmikosR — APPROVED** [review]

**Comment by XhmikosR** [comment]
Please wait for @mdo's feedback too.

**Comment by mdo** [comment]
This wouldn't be a patch release, this would be a minor or major as it adds new functionality IMO. Could be convinced otherwise though.

That aside, does anything change once we've removed the jQuery dependency? Does this gain us anything long term? Is there a path forward without generating more dist files?

**Comment by Johann-S** [comment]
Nope nothing will change about that when jQuery will be removed, I choose to build our plugins in UMD (our other dist files are in UMD too, except `js/dist`) because it's compatible with most of bundlers so we'll keep those file for a long time.

Unfortunately there is no other path, for example Popper.js (see https://cdnjs.com/libraries/popper.js) is built in three format (UMD, ESM and IIFE)

**Comment by stof** [comment]
The issue is that even without jQuery, your plugins still depend on Util, and these internal dependencies are incompatible with bundlers if they rely on global variables all the time.

One solution could be to turn the existing `js/dist` into UMD ones all the time. It would still work for people using them globally (that's the point of UMD: it works for both global-based loading and CommonJS)

**Comment by Johann-S** [comment]
Unfortunately it's doesn't work if we build our `js/dist` files in UMD, because the scope of `Util` won't be reachable for our other plugins, that's why I let them in IIFE

**Comment by stof** [comment]
@Johann-S that's not true. UMD builds can still expose their export to be consumed by others. That's the whole point of UMD.

**Comment by Johann-S** [comment]
I tried what you said @stof and that's what I think too, but that's not what I get... Maybe I miss something in my Rollup config 🤔

**Comment by Johann-S** [comment]
Finally I found a solution using only Babel without Rollup.

But for those who do not use a bundler, to use our plugins they will have to add `['default']` for example: 
```js
Util['default'].getUid('test')
```

For me it's a breaking change, but with this solution no new dist files... Any feedbacks ?
 
/CC @XhmikosR and @mdo

**Comment by mdo** [comment]
@Johann-S I don't know enough about it at the moment to make a call myself. Might need to ask you to explain more over Slack?

**Comment by Johann-S** [comment]
Sure @mdo will talk about it on Slack 👌

**Comment by Johann-S** [comment]
Hi everybody !

Finally I made a change in my script, and...
- No new dist files
- No breaking change
- One dev deps nuked

So IMO, we should merge that asap.

/CC @XhmikosR if you can check my docs change 😄 , and @mdo to notice you

**Comment by donni106** [comment]
@Johann-S how does jquery work for you? i always get TypeError: $ is undefined (referenced in util.js:64:5)

```
import $ from 'jquery';
import 'bootstrap/js/dist/util';
import 'bootstrap/js/dist/carousel';
```

my project is setup with a rollup config where i already tried a thousand things for making jquery work :(

**Comment by Johann-S** [comment]
Pretty difficult to help you without seeing your Rollup config @donni106

**Comment by donni106** [comment]
thanks. ok, currently it is the following. but i tried with setting `globals` or `commonjs` stuff before. i read a few things but nothing made any changes.

```
const path = require('path');
const babel = require('rollup-plugin-babel');
const resolve = require('rollup-plugin-node-resolve');
const commonjs = require('rollup-plugin-commonjs');

const fileDest = 'script.js';

module.exports = {
  input: path.resolve(__dirname, '../src/tx_theme/js/main.js'),
  output: {
    banner: '',
    file: path.resolve(
      __dirname,
      `../dist/typo3/typo3conf/ext/theme/Resources/Public/JS/${fileDest}`
    ),
    format: 'umd',
    globals: {},
    name: 'script',
  },
  external: ['jquery'],
  plugins: [
    babel({
      exclude: 'node_modules/**',
    }),
    resolve(),
    commonjs(),
  ],
};
```

**Pull request merged**